File: examples/python/output/pager.py
#!/usr/bin/env python3
"""Route long output through $PAGER / `less -R`.

    make run-example EX=python/output/pager

In a terminal the table opens in the pager (press `q`). Off a TTY the pager
is skipped and the output prints directly.
"""

from __future__ import annotations

import sparcli as sc

ROW_COUNT = 100


def main() -> None:
    # Everything inside the block is paged; exit_status is set on exit.
    with sc.Pager() as pager:
        sc.rule('100 rows', sc.RuleOpts(type=sc.BorderType.DOUBLE))

        table = sc.Table()
        table.column('#', sc.ColOpts(halign=sc.Align.RIGHT)) \
             .column('Square', sc.ColOpts(halign=sc.Align.RIGHT))
        for i in range(1, ROW_COUNT + 1):
            table.row([str(i), str(i * i)])
        table.print(sc.TableOpts(border=sc.BorderType.SINGLE, header_row=True,
                                 header_style=sc.Style.bold()))

    print(f'pager exit status: {pager.exit_status}')


if __name__ == '__main__':
    main()

File: examples/rust/output/pager.rs
//! Route long output through $PAGER / `less -R`.
//!
//!   cargo run -p sparcli --example output_pager
//!
//! In a terminal the table opens in the pager (press `q`). Off a TTY the
//! pager is skipped and the output prints directly.

use sparcli::*;

const ROW_COUNT: i32 = 100;

fn main() {
    // Everything until end()/drop is paged.
    let pager = Pager::begin(PagerOpts::new());

    rule(RuleOpts::new().kind(BorderType::Double).title("100 rows"));

    let mut table = Table::new();
    table
        .column("#", ColOpts::new().align(Align::Right))
        .column("Square", ColOpts::new().align(Align::Right));
    for i in 1..=ROW_COUNT {
        table.row([i.to_string(), (i * i).to_string()]);
    }
    table.print(TableOpts::new().border(BorderType::Single).header_row(true));

    let status = pager.end();
    println(&format!("pager exit status: {status}"), Style::default());
}
